api/xds_protocol.rst

```diff
@@ -155,45 +231,109 @@ the resources in the :ref:`DiscoveryResponse <envoy_api_msg_DiscoveryResponse>`
-server to spin and have a severe performance impact.
+waiting for a change to occur, it will cause needless work on both the client and the management
+server, which could have a severe performance impact.
 
 Within a stream, new :ref:`DiscoveryRequests <envoy_api_msg_DiscoveryRequest>` supersede any prior
 :ref:`DiscoveryRequests <envoy_api_msg_DiscoveryRequest>` having the same resource type. This means that
 the management server only needs to respond to the latest
 :ref:`DiscoveryRequest <envoy_api_msg_DiscoveryRequest>` on each stream for any given resource type.
 
-Resource hints
-^^^^^^^^^^^^^^
+How the client specifies what resources to return
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+xDS requests allow the client to specify a set of resource names as a hint to the server about which resources the client
+is interested in. In the SotW API variants, this is done via the :ref:`resource_names
+<envoy_api_field_DiscoveryRequest.resource_names>` specified in the :ref:`DiscoveryRequest
+<envoy_api_msg_DiscoveryRequest>`; in the incremental API variants, this is done via the
+:ref:`resource_names_subscribe <envoy_api_field_DeltaDiscoveryRequest.resource_names_subscribe>` and
+:ref:`resource_names_unsubscribe <envoy_api_field_DeltaDiscoveryRequest.resource_names_unsubscribe>` fields in the
+:ref:`DeltaDiscoveryRequest <envoy_api_msg_DeltaDiscoveryRequest>`.
 
-The :ref:`resource_names <envoy_api_field_DiscoveryRequest.resource_names>` specified in the :ref:`DiscoveryRequest
-<envoy_api_msg_DiscoveryRequest>` are a hint about what resources the client is interested in.
+For `Listener` and `Cluster` resource types, Envoy will always set the resource hints to empty, in which case the server
+should use site-specific business logic to determine the full set of resources that the client is interested in, typically
+based on the client's :ref:`node <envoy_api_msg_Core.Node>` identification. However, other xDS clients (such as gRPC) may
+specify explicit LDS/CDS resources as resource hints, for example if they only have a singleton listener and already know
+its name from some out-of-band configuration.
 
-For :ref:`Listener Discovery Service (LDS) <envoy_api_msg_Listener>` and :ref:`Cluster Discovery Service (CDS)
-<envoy_api_msg_Cluster>`, Envoy will always set the resource hints to empty, in which case the server should use
-site-specific business logic to determine the full set of resources that the client is interested in, typically based on
-the client's node identification. However, other xDS clients may specify explicit LDS/CDS resources as resource hints, for
-example if they only have a singleton listener and already know its name from some out-of-band configuration. For EDS/RDS, the
-resource hints are required.
+For other resource types, the resource hints are required. If the field is empty, that means that the client is not
+interested in any resources of the relevant type.
```

Comment:
So `resource_names` being empty has almost opposite meanings between Listener/Cluster and other resource types. I understand the Envoy implementation dictates this spec to some extent (and this is Envoy doc we are proposing changes to) but the inconsistency is a bit concerning and I wonder if it can be fixed.

Reply:
I agree that this is inconsistent and confusing, but I don't think there's a way to fix it without breaking backward compatibility.

Ultimately, I think the best way to make this problem go away is to try to migrate to the incremental API, in which all resource types are handled the same way.

Reply:
Actually, I guess the incremental API doesn't really solve this problem.  I'll let @htuch and @mattklein123 weigh in on whether this is something we can fix in xDS or whether it will need to wait for UDPA.  But even if it can be fixed in xDS, I think that would have to be done separately from this PR.

Reply:
If Envoy never uses empty `resource_names` for EDS/RDS (*other resource types*), we should be able to define that behavior here to be more consistent. But I am not sure if the assumption is true and if it is a worthwhile exercise.

Reply:
I believe that the behavior in that case is already defined here.  Do you see some part of it that behavior that you don't think is covered here?

Reply:
What I meant to say is to change that definition (of the behavior) to make it consistent between the 2 cases. Specifically in both cases (Listener/Cluster *AND* other resource types) when the client has empty `resource_names` in the request the management server sends "state of the world" for that resource type (or whatever business logic it wants to use to figure out resources the client might be interested in).

This change of definition will only work if Envoy never uses empty `resource_names` for other resource types and that might be an incorrect assumption. Also not sure about the benefit of making this behavior consistent.

Reply:
I think that change would be strictly worse than the status quo.  Currently, the semantics for RDS and EDS allow unsubscribing to the last resource, but the semantics for LDS and CDS do not.  The change you're proposing would make it impossible to unsubscribe to the last resource in RDS and EDS as well.  In effect, this would be taking the case that *does* work and breaking it, for no real benefit.

Reply:
Yes, agreed.

Reply:
I think we've now addressed the unsubscription case below.